banlist.dat: store banlist on disk

jonasschnelli opened the pull request:
Currently, the list of banned nodes are only held in memory. A restart of bitcoind would reset the set of banned nodes.

This PR introduces a new filestore (`banlist.dat`) which stores the IPs/Subnets of the banned nodes more or less identical to the CAddrDB (`peers.dat`).

Also added within this PR is a feature that removes unused banned node entries because the `banned_till` time has expired.

includes tests for the banlist disk storage.

Unsure if we should introduce a new cmd arg `-storebanlist` to optionally allow storing of banned node data. It could be, that some users expecting a sweep of all banned nodes when they restart bitcoind.

jonasschnelli committed:
banlist.dat: store banlist on disk

jgarzik commented:
Nice feature.  RE -storebanlist, the system should default to storing.   Users may disable.

Or don't create an option at all -- same as CAddrDB, user may delete file to clear the list.

Diapolo commented:
I'd also vote for treating this the same as peers.dat and don't add an option for it.

Diapolo commented on src/net.cpp:
```diff
@@ -2107,3 +2173,119 @@ void CNode::EndMessage() UNLOCK_FUNCTION(cs_vSend)
+        return error("%s: Failed to open file %s", __func__, pathBanlist.string());
+
+    // use file size to size memory buffer
+    int fileSize = boost::filesystem::file_size(pathBanlist);
```
Not sure if this is best to use an int here... what is file_size returning?

jonasschnelli replied:
Do you expect users with more then 2GB of banned node data?
I have taken this approach from CAddrDB 1:1. But i agree file sizes should be held in `uint64_t` or in `size_t` if they are not serialized/shared anywhere.

Diapolo replied:
I think it's plenty of space, but we should be 100% sure we don't overflow IMHO? Yeah there are more parts in the code that don't do that.

laanwj replied:
Right: to be precise, file sizes should be `uint64_t`, not `size_t`. Remember that size_t depends on the address width, it is for in-memory sizes.

jonasschnelli replied:
Agreed. Changed from `int` to `uint64_t` (also in `CAddDB`).

Diapolo commented on src/net.cpp:
```diff
@@ -2107,3 +2173,119 @@ void CNode::EndMessage() UNLOCK_FUNCTION(cs_vSend)
+    // Don't try to resize to a negative number if file is small
+    if (dataSize < 0)
+        dataSize = 0;
+    vector<unsigned char> vchData;
```
Maybe you could use std:: so we can get rid of using namespace std; easier :)?

dexX7 commented:
> @jgarzik: RE -storebanlist, the system should default to storing. Users may disable.

This seems risky in my opinion, because not all IPs are static, and they don't necessarily always belong to the same node.

sipa commented:
dexX7: that is why all bans are temporary.

jonasschnelli: I think defaulting to storing them is perfectly fine (or
even not offering the option).

jonasschnelli commented:
@jgarziks way of resetting the banlist (by removing banlist.dat) is very straightforward and doesn't need any explanation. I'd like to avoid another cmd arg option to not end up in having thousands of tiny options.

Node misbehaving bans are by default 24h. All other bans needs conscious actions from the users.
The chance that your node accidentally get an IP out of a dynamic range, where the previous owner managed to add his bitcoin-core on a ban-list, is very rar. And then, the chance that all nodes did ban this IP or that you connect to a node which conscious added your new IP to the banlist is even rarer
We have 3,706,452,992 assignable public IPv4 addresses.

laanwj commented:
Concept ACK, also agree that defaulting to store bans is perfectly fine (so is not offering an option).

As any kind of persistence commits to a format it is good to think forward a bit what should be included: would it make sense to add e.g. a "ban source" enum field, that specifies whether the ban is automatic or manual?

jonasschnelli committed:
CAddrDB/CBanDB: change filesize variables from int to uint64_t

jonasschnelli commented:
I like the idea of storing the "ban source". Because we now keep/dump a serialized `std::map<CSubNet, int64_t>` everything beyond that would be a bigger change. We should probably introduce a simple class for the banlist entry like `class CBanEntry { int64_t bantime, enum banSource }`. This might also be capable of allowing ban distinction between multiple bans on the same IP but with different ports (but would mean moving from std::map to a vector).

laanwj commented:
> We should probably introduce a simple class for the banlist entry like class CBanEntry { int64_t bantime, enum banSource }

Right. That would also allow for versioning in serialization.

> This might also be capable of allowing ban distinction between multiple bans on the same IP but with different ports

This might be overdoing it; at least I don't see much added value in per-port bans, at the least they're easily circumvented.

Diapolo commented:
@laanwj Think about per port bans with incoming Tor/proxy connections, they're all 127.0.0.1 and when I currently ban a single 127.0.0.1 with a specific port all further connection attempts are blocked until bantime is over. Not sure if this is to solve in any way.

laanwj commented:
@Diapolo That simply won't work. As I try to explain above, source ports aren't unique nor identifying.

Diapolo commented:
@laanwj Indeed, my fault... but should be warned when banning 127.0.0.1 or do you assume people know what they are banning?

jonasschnelli committed:
use CBanEntry as object container for banned nodes

- added a reason enum for a ban
- added creation time for a ban

Using CBanEntry as container will keep banlist.dat extenable.

jonasschnelli commented:
Followed and implemented @laanwj s proposal.
Added `CBanEntry` as ban metadata container which uses `enum BanReason` (`BanReasonNodeMisbehaving`, `BanReasonManuallyAdded`). Also added `nCreateTime` within the new metadata class, to allow to backtrack when a ban was added/created.
Using now everywhere the typedef `std::map<CSubNet, CBanEntry> banmap_t`.

Now the `banlist.dat` file is extendable over the possibility to detect different versions (`CBanEntry->nVersion`) of entries and allow possible migrations during deserialization.

laanwj commented on src/rpcnet.cpp:
```diff
@@ -541,15 +542,19 @@ UniValue listbanned(const UniValue& params, bool fHelp)
         UniValue rec(UniValue::VOBJ);
         rec.push_back(Pair("address", (*it).first.ToString()));
-        rec.push_back(Pair("banned_untill", (*it).second));
+        rec.push_back(Pair("banned_untill", banEntry.nBanUntil));
```
`s/untill/until/` I suppose?

jonasschnelli replied:
> s/untill/until/ I suppose?

Fixed typo.

laanwj commented:
utACK

laanwj commented on src/net.cpp:
```diff
@@ -474,30 +476,35 @@ bool CNode::IsBanned(CSubNet subnet)
 
-void CNode::Ban(const CNetAddr& addr, int64_t bantimeoffset, bool sinceUnixEpoch) {
+void CNode::Ban(const CNetAddr& addr, const BanReason &banReason, int64_t bantimeoffset, bool sinceUnixEpoch) {
     CSubNet subNet(addr.ToString()+(addr.IsIPv4() ? "/32" : "/128"));
```
Hhmm at some point we should add a CSubNet constructor or factory function that makes a subnet with one unique `CNetAddr` (netmask all ones), and use that, instead of going through a string. Not a big deal, though.

jonasschnelli replied:
> Hhmm at some point we should add a CSubNet constructor or factory function that makes a subnet with one unique CNetAddr (netmask all ones), and use that, instead of going through a string. Not a big deal, though.

Right. Added (see latest commit). If we don't add it now, it probably will never be added.

laanwj commented on src/net.cpp:
```diff
@@ -474,30 +476,35 @@ bool CNode::IsBanned(CSubNet subnet)
+    banEntry.nBanUntil = GetTime()+GetArg("-bantime", 60*60*24);  // Default 24-hour ban
     if (bantimeoffset > 0)
-        banTime = (sinceUnixEpoch ? 0 : GetTime() )+bantimeoffset;
+        banEntry.nBanUntil = (sinceUnixEpoch ? 0 : GetTime() )+bantimeoffset;
```
maybe
```
if (bantimeoffset <= 0)
    bantimeoffset = GetArg("-bantime", 60*60*24); // Default 24-hour ban;
banEntry.nBanUntil = (sinceUnixEpoch ? 0 : GetTime() )+bantimeoffset;    
```
(slightly cleaner, not calling GetTime() twice)

jonasschnelli replied:
Yes. This makes more sense. Fixed and pushed.

ajweiss commented:
I have an unsubmitted patch that also adds "remembering" of ban scores associated by IP address across disconnects over a time window (24h I think).  That is, it prevents malicious nodes from carrying out sawtooth attacks where they can reset their ban score by disconnecting and reconnecting before being banned.  There are pros (prevents malicious behavior) and cons (tightens up the rules and increases the likelihood of NAT exit hosts getting banned due to lots of small accumulations of ban points over time.)  Is there any interest in this?  I can open a new pull if so...

jonasschnelli committed:
Adding CSubNet constructor over a single CNetAddr

laanwj commented:
@ajweiss Not sure about that. At least make sure that the associated structure is size-limited. Especially with IPv6 it will be extremely easy to fill memory that way. But this needs to be  discussed somewhere else as it is not related to this pull.

laanwj commented on src/netbase.cpp:
```diff
@@ -1291,6 +1291,24 @@ CSubNet::CSubNet(const std::string &strSubnet, bool fAllowLookup)
+    memset(netmask, 255, sizeof(netmask));
+
+    std::vector<CNetAddr> vIP;
+    if (!LookupHost(addr.ToString().c_str(), vIP, 1, fAllowLookup))
```
Thanks. But this seems a bit circuitous. Let's do just:
```
memset(netmask, 255, sizeof(netmask));
network = addr;
```
Also the fAllowLookup parameter is unnecessary. It should never be needed to look up anything when going from an already binary address.

jonasschnelli replied:
Right. Wasn't aware of the `LookupHost()` in the `CNetAddr::CNetAddr()` constructor.
Fixed and pushed.

jonasschnelli replied:
Right! The isValid must be set explicit. Fixed.
